packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts

```diff
@@ -775,6 +793,168 @@ export class SafeDsPythonGenerator {
+            expression,
+        )`${RUNNER_SERVER_PIPELINE_MANAGER_PACKAGE}.runner_memoized_function_call("${fullyQualifiedTargetName}", ${
+            containsOptionalArgs ? 'lambda *_ : ' : ''
+        }${
+            containsOptionalArgs
+                ? this.generatePlainCall(expression, sortedArgs, frame)
+                : isSdsMemberAccess(expression.receiver) && isSdsCall(expression.receiver.receiver)
+                  ? expandTracedToNode(expression.receiver)`${this.generateExpression(
+                        expression.receiver.receiver.receiver,
+                        frame,
+                    )}.${this.generateExpression(expression.receiver.member!, frame)}`
+                  : this.generateExpression(expression.receiver, frame)
+        }, [${thisParam ? thisParam.append(', ') : ''}${joinTracedToNode(expression.argumentList, 'arguments')(
+            memoizedArgs,
+            (arg) => this.generateExpression(arg, frame),
+            {
+                separator: ', ',
+            },
+        )}], [${joinToNode(hiddenParameters, (param) => param, { separator: ', ' })}])`;
+    }
+
+    private getMemoizedCallHiddenParameters(expression: SdsCall, frame: GenerationInfoFrame): CompositeGeneratorNode[] {
+        const impurityReasons = this.purityComputer.getImpurityReasonsForExpression(expression);
+        const hiddenParameters: CompositeGeneratorNode[] = [];
+        for (const reason of impurityReasons) {
+            if (reason instanceof FileRead) {
+                if (typeof reason.path === 'string') {
+                    hiddenParameters.push(
+                        expandTracedToNode(
+                            expression,
+                        )`${RUNNER_SERVER_PIPELINE_MANAGER_PACKAGE}.runner_filemtime('${reason.path}')`,
+                    );
+                } else if (isSdsParameter(reason.path)) {
+                    const argument = this.nodeMapper
+                        .parametersToArguments([reason.path], expression.argumentList.arguments)
+                        .get(reason.path);
+                    if (!argument) {
+                        /* c8 ignore next 4 */
+                        throw new Error(
+                            'File Read impurity with dependency on parameter is present on call, but no argument has been provided.',
+                        );
+                    }
+                    hiddenParameters.push(
+                        expandTracedToNode(
+                            argument,
+                        )`${RUNNER_SERVER_PIPELINE_MANAGER_PACKAGE}.runner_filemtime(${this.generateArgument(
+                            argument,
+                            frame,
+                        )})`,
+                    );
+                }
+            }
+        }
+        return hiddenParameters;
+    }
+
+    private getFullyQualifiedReferenceName(expression: SdsExpression, declaration: SdsDeclaration): string | undefined {
```

Comment:
I don't understand this method. 

You always return the equivalent of
`this.builtinAnnotations.getPythonModule(targetModule) || targetModule.name}.${this.getPythonNameOrDefault(declaration)`, unless the declaration is in a different module and not imported explicitly. In the latter case you return undefined.

Why does it make a difference whether the declaration is imported explicitly or not? For instance, everything in a `safeds.xy` package is available without imports, but the notion of a qualified name still makes sense.

At first glance it seems like you can remove this method completely and use `getQualifiedName` instead.

Reply:
Thanks for the hint, that makes it a lot easier. I cannot use `getQualifiedName` directly because of python module/python name differences.

Reply:
Right, that makes sense.